backend/main.go

```diff
@@ -42,20 +45,61 @@ const (
 type Server struct {
 	store        storage.StorageService
 	storeTimeout time.Duration
+	auth         Auth
+}
+
+type Auth interface {
+	authenticate(next http.HandlerFunc) http.HandlerFunc
+}
+
+type noAuth struct{}
+
+func (auth noAuth) authenticate(next http.HandlerFunc) http.HandlerFunc {
+	return next
+}
+
+type firebaseAuth struct {
+	firebaseClient authClient
+}
+
+type authClient interface {
+	VerifyIDToken(ctx context.Context, idToken string) (*auth.Token, error)
+}
+
+func (auth firebaseAuth) authenticate(next http.HandlerFunc) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		ctx, cancel := context.WithTimeout(r.Context(), defaultStoreTimeout)
+		defer cancel()
+
+		idToken := strings.TrimPrefix(r.Header.Get("Authentication"), "Bearer ")
+		_, err := auth.firebaseClient.VerifyIDToken(ctx, idToken)
+		if err != nil {
+			log.Printf("User authentication failed: error: %s", err)
+			http.Error(w, "User authentication failed (see server log)", http.StatusUnauthorized)
+			return
+		}
+
+		next(w, r)
+	}
+}
+
+func getDomain(email string) string {
```

Comment:
I think this is part of your later authorization-related change, right? I don't see it called anywhere here. It probably shouldn't be part of this pull request.

Reply:
Yep, that's true. I've removed it.